### PR TITLE
Docs: Add snapshot-testing doc

### PR DESCRIPTION
**Pull request by sirreal**
This PR adds a snapshot testing doc based on p4TIVU-8Hv-p2

Related: #19882

### CODE REVIEW

File: docs/testing/snapshot-testing.md [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+```
+
+You should never interact with the snapshot directly, they are generated by tests. However you can see that it's a representation the test data.
+
+## Advantages
+
+
+* Trivial and concise to write tests
+* Protect against unintentional changes
+* Simple to work with
```
Comment (gziolo) [inline docs/testing/snapshot-testing.md]: I figured out that it is an amazing way to see the representation of the component without loading the browser.

File: docs/testing/snapshot-testing.md [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+  test( 'should render with locale suggestions if locale is provided', () => {
+    const wrapper = shallow( <MyComponent locale="es" /> );
+    expect( wrapper ).toMatchSnapshot();
+    expect( wrapper.find( LocaleSuggestions ) ).toHaveLength( 1 );
+  } );
+} );
+```
+
+But the first snapshot tests introduced in Calypso were actually <a href="https://github.com/Automattic/wp-calypso/blob/e34d15f44c261fd7daa2212017e995883866d603/client/state/comments/test/selectors.js#L133-L142">reducer tests</a>. Brilliant! When we work with reducers we're often working with large data structures that we don't want to change unexpectedly.
```
Comment (gziolo) [inline docs/testing/snapshot-testing.md]: Why not [reducer tests](https://github.com/Automattic/wp-calypso/blob/e34d15f44c261fd7daa2212017e995883866d603/client/state/comments/test/selectors.js#L133-L142)? Asking out of my own curiosity.

Comment (sirreal) [reply docs/testing/snapshot-testing.md]: 🤔  Apparently this doesn't read well, this is supposed to say:

> Reducer tests are also great fit for snapshots.

I'll adjust for clarity.

**Comment by gziolo** [comment]
This is awesome. I think it's covered in-depth and explains all the core concepts. This is how we should be treating snapshot testing inside Calypso.

Now the question is how we can connect it with the existing documents to make it better exposed. We should at least reference this document in this section: https://github.com/Automattic/wp-calypso/blob/fd01ab6d245c5be3a0a7f329cdd7557f08ee8081/docs/testing/testing-overview.md#client-side-tests.

File: docs/testing/snapshot-testing.md [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+When a snapshot test fails, it just means that a component's rendering has changed. If that was unintended,  then the snapshot test just prevented a bug 😊
+
+However, if the change was intentional, follow these steps to update the snapshot:
+
+1. Run the following to update the snapshots:
+   ```sh
+   # --testPathPattern is optional but will be much faster by only running matching tests
+   npm run test-client -- --updateSnapshot --testPathPattern client/components
+   ```
+1. Review the diff and ensure the changes are expecte and intentional
```
Comment (gziolo) [inline docs/testing/snapshot-testing.md]: Nice trick with numbers :)

File: docs/testing/snapshot-testing.md (hunk 2 above)
Comment (gziolo) [reply docs/testing/snapshot-testing.md]: My question was about the syntax used 😛 

On the second thought, your comment makes sense. Better to give a context for external contributors.

File: docs/testing/snapshot-testing.md (hunk 3 above)
Comment (mtias) [reply docs/testing/snapshot-testing.md]: Type on `expecte`

File: docs/testing/snapshot-testing.md (hunk 1 above)
Comment (sirreal) [reply docs/testing/snapshot-testing.md]: * Representation of the component without needing the DOM

Something like that?

Comment (gziolo) [reply docs/testing/snapshot-testing.md]: Maybe broader:

> Reveal the internal structure without need to run the application

File: docs/testing/snapshot-testing.md (hunk 2 above)
Comment (sirreal) [reply docs/testing/snapshot-testing.md]: I went with

> Reducer tests are also be a great fit for snapshots. The firsts snapshots introduced in Calypso were actually reducer tests. Reducers can be large, complex data structures that we don't want to change unexpectedly, exactly what snapshots excel at!

**Pull request merged**